File: 프로그래머스/3/43164. 여행경로/여행경로.py
from collections import defaultdict

def solution(tickets):
    answer: list[str] = []
    route: defaultdict[str, list[str]] = defaultdict(list)
    
    for ticket in tickets:
        start: str
        end: str
        start, end = ticket
        route[start].append(end)
    
    for key in route:
        route[key].sort(reverse=True)
    
    stack: list[str] = ["ICN"]
    while stack:
        top: str = stack[-1]
        if route[top]:
            stack.append(route[top].pop())
        else:
            answer.append(stack.pop())
    
    return answer[::-1]

File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;

class Solution {
    public String[] solution(String[][] tickets) {
        Map<String, List<String>> map = new HashMap<>();
        for (String[] ticket : tickets) {
            String from = ticket[0];
            String to = ticket[1];
            if (!map.containsKey(from)) map.put(from, new ArrayList<>());
            map.get(from).add(to);
        }
        for (String key : map.keySet()) Collections.sort(map.get(key));

        List<String> route = new ArrayList<>();
        Stack<String> stack = new Stack<>();
        stack.push("ICN");
        
        while (!stack.isEmpty()) {
            String city = stack.peek();
            if (map.containsKey(city) && !map.get(city).isEmpty()) {
                String next = map.get(city).remove(0);
                stack.push(next);
            } else {
                route.add(0, stack.pop());
            }
        }
        
        return route.toArray(new String[route.size()]);
    }
}